### PR TITLE
[ART-1353] Add script to identify buildroot of golang packages

### PR DESCRIPTION
**Pull request by thiagoalessio**
Adding this script to aos-cd-jobs, so we can either use it standalone or call
it from our other tools (doozer/elliott)

```
$ ./golang-built-rpms -h
golang-built-rpms v0.0.1

Find golang-built RPMs and print out their corresponding buildroots.

Usage:
    $ golang-built-rpms OCP_VERSION [OPTIONS]...

Options:
    -f    --format     Choose an output format.
                       Available formats: csv yaml
                       default: csv

    -v    --verbose    Print skipped packages to STDERR
                       default: 0

    -h    --help       Print this help

Examples:
    Most basic usage:
    $ golang-built-rpms 3.11

    Logging skipped (non-golang) packages to file:
    $ golang-built-rpms 4.1 -v 2> non-golang-pkgs.log

    YAML output to file
    $ golang-built-rpms 4.2 -f yaml > pkg-buildroots.yaml
```

```
$ time ./golang-built-rpms 4.2 -v -f csv >4.2-golang-built-rpms.csv 2>skipped.log

real    4m34.810s
user    10m26.645s
sys     3m43.020s
```
Output files from command above are attached to https://issues.redhat.com/browse/ART-1353

### CODE REVIEW

**Comment by thiagoalessio** [comment]
Just realized that `-v` `--verbose` is not making any difference ... fixing it now

File: build-scripts/golang-built-rpms/golang-built-rpms [hunk 1]
```diff
@@ -0,0 +1,302 @@
+#!/usr/bin/env bash
+
```
Comment (vfreex) [inline build-scripts/golang-built-rpms/golang-built-rpms]: Needs `set -euo pipeline` to prevent silent failure on exception.

Comment (sosiouxme) [reply build-scripts/golang-built-rpms/golang-built-rpms]: agreed, this should pretty much come standard (even when hacking up a script...)

**Comment by thiagoalessio** [comment]
/hold

We intend to discuss a potential new home for this tool on today's arch meeting.

**Comment by thiagoalessio** [comment]
WIP: testing latest changes
* limited concurrency to 1000 parallel processes (it was freezing my computer everytime)
* trying to fix multiple "latest_candidates" (1 for rhel-7 and 1 for rhel-8) ... my code was assuming only one would be returned at all times

**Comment by thiagoalessio** [comment]
fixed the issues I found, I'm already making use of it for [ART-1547](https://issues.redhat.com/browse/ART-1547).
There is one small detail though ... I commented `latest_build` because I wasn't use it to get any further info, and seems a pain to map multiple `latest_builds` with their corresponding `latest_candidates` ... I'd like to discuss with y'all what do we loose by not having this field.

**Comment by thiagoalessio** [comment]
one more thing ... i was really expecting `podman` and `cri-o` to be picked on 4.2 ... (they are recognized as golang RPMs on 4.3, but got skipped on 4.2)
is there any outside reason for that, or my script is just flawed ?

File: build-scripts/golang-built-rpms/golang-built-rpms (hunk 1 above)
Comment (thiagoalessio) [reply build-scripts/golang-built-rpms/golang-built-rpms]: :man_facepalming: of course ... resolved.

**Comment by sosiouxme** [comment]
> they are recognized as golang RPMs on 4.3, but got skipped on 4.2

the el7 version of cri-o isn't skipped. the el8 version is because:

```
🐚 rpm -qp --requires http://download.eng.bos.redhat.com/brewroot/vol/rhel-8/packages/cri-o/1.13.12/6.dev.rhaos4.1.git8abaaeb.el8_0/src/cri-o-1.13.12-6.dev.rhaos4.1.git8abaaeb.el8_0.src.rpm
compiler(go-compiler)
git
glib2-devel
glibc-static
go-md2man
gpgme-devel
libassuan-devel
libseccomp-devel
pkgconfig(systemd)
rpmlib(CompressedFileNames) <= 3.0.4-1
rpmlib(FileDigests) <= 4.6.0-1
```
no golang in there :-) have to look for go-compiler too apparently.
I found a couple other things to tweak. Will add a commit and merge.

**Comment by sosiouxme** [comment]
one other note: looking up latest builds individually is pretty slow (lot of brew calls, each has to negotiate tls and auth...).

you could do this all at once with ```
brew list-tagged --latest --inherit $tag
```

remove V and R to get N, then set up an associative array and you have an instant lookup (although slightly tricky since there could be two entries).

you might even be able to filter out containers at the same time (not have to look up their buildinfo). they are inherited from a different tag when they're shipped, e.g.

```
$ brew list-tagged --latest --inherit rhaos-3.11-rhel-7
search-guard-ssl-2.4.4.19_redhat_1-1.el7  rhaos-3.7-rhel-7      rmeggins
snapshot-controller-container-v3.11.170-5  rhaos-3.11-rhel-7-container-released  ocp-build/buildvm.openshift.eng.bos.redhat.com
[...]
```

regarding latest shipped vs latest candidate... let's see how this shakes out when used in earnest. we are likely interested in both (what built what we've shipped, and do our candidate builds fill the gap?).

**Pull request merged**